apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -133,35 +211,158 @@ type ReadinessCheck struct {
-// transformers.
+// patchSetMap stores a list of patches from a set, against the Set ID.
+type patchSetMap map[string]Patch
+
+// PatchType defines the usable inbuilt patch types
+type PatchType string
+
+// Predefined PatchTypes.
+const (
+	PatchTypeFromCompositeFieldPath PatchType = "FromCompositeFieldPath" // Default
+	PatchTypePatchSet               PatchType = "PatchSet"
+)
+
+// Patch instances are applied between composite and composed resources.
+// Their behaviour depends on the Type value selected.
 type Patch struct {
+	// Type sets the patching behaviour to be used. Each patch type may require
+	// its' own fields to be set on the Patch object.
+	// +optional
+	// +kubebuilder:validation:Enum=FromCompositeFieldPath;PatchSet
+	// +kubebuilder:default=FromCompositeFieldPath
+	Type PatchType `json:"type,omitempty"`
+
+	// PatchSet allows other patch types to be nested under itself.
+	// PatchSets are defined on a Composition and can be referenced here,
+	// using their ID.
+	// +optional
+	*PatchSet `json:",inline"`
+
+	// PatchFromCompositeFieldPath is the 'standard' patch type that maps a field
+	// path from the composite resource to the composed resource, with optional
+	// transformations.
+	// +optional
+	*PatchFromCompositeFieldPath `json:",inline"`
+}
 
+// Apply applies a patch between the from and to objects. Depending on the
+// patch type, this may be from the composite resource to the composed resource,
+// or vice-versa.
+func (c *Patch) Apply(from, to runtime.Object) error {
+	var patcher interface {
+		Apply(from, to runtime.Object) error
+	}
+
+	switch c.Type {
+	case PatchTypeFromCompositeFieldPath:
+		patcher = c.PatchFromCompositeFieldPath
+	case PatchTypePatchSet:
+		patcher = c.PatchSet
+	default:
+		return errors.Errorf(errInvalidPatchType, c.Type)
+	}
+
+	return patcher.Apply(from, to)
+}
+
+// PatchSetCompatiblePatch embeds the valid PatchTypes which can be nested within
+// a PatchSet.
+type PatchSetCompatiblePatch struct {
+	// Type sets the patching behaviour to be used. Each patch type may require
+	// its' own fields to be set on the Patch object.
+	// +optional
+	// +kubebuilder:validation:Enum=FromCompositeFieldPath;PatchSet
+	// +kubebuilder:default=FromCompositeFieldPath
+	Type PatchType `json:"type,omitempty"`
+
+	// PatchFromCompositeFieldPath is the 'standard' patch type that maps a field
+	// path from the composite resource to the composed resource, with optional
+	// transformations.
+	// +optional
+	*PatchFromCompositeFieldPath `json:",inline"`
+}
+
+// PatchSet defines a list of patches as a single grouping.
+// It can contain all other patch types except itself.
+type PatchSet struct {
+	// ID is the unique identifier of the PatchSet. ID is required when
+	// defining a new PatchSet, and also refers to a PatchSet from within
+	// the list of Patches on a resource.
+	// +optional
+	ID *string `json:"id,omitempty"`
```

Comment:
I'm afraid that per our API sketch at https://github.com/crossplane/crossplane/issues/1974#issuecomment-733146966 @muvaf convinced me that `name` was preferable to `id` here after all. Would you mind updating?

We'd also like to call this `patchSetName` when referring to a patchset from a resource's array of patches to make it very clear what the field is for, so we probably can't use a single type for both cases.

Reply:
I'm all aboard with the API described in #1974 - however I've been looking at implementing this now as 2 different types, and I'm having trouble getting my head around how to allow this to work without having extraneous fields that appear in the CRD definition.

Basically, if we define a `PatchSetDefinition` on the Composition, at some point that `PatchSetDefinition` has to be 'copied' to <somewhere> on the `ComposedTemplate` to allow it to be defined at some point in the future. Using the one `PatchSet` type which was embedded into `Patch`, it was possible to simply create a new `Patch` with the `PatchSet` struct configured.

With the additional type to allow the Name field to be changed in JSON (which has its' own benefits from a validation perspective), if we embed both into `Patch` then we end up with the following field definitions in the CRD:

```yaml
resources:
    items:
      properties:
        ...
        patches:
          items:
            properties:
              # Defined on CompositeFieldPath
              fromFieldPath:
                type: string
              # Defined on PatchSetDefinition
              name:
                description: Name is the unique identifier of the PatchSet. It is required when defining a new PatchSet, and used to refer to a PatchSet from within the list of Patches on a resource.
                type: string
              # Defined on PatchSet
              patchSetName:
                description: Name is the unique identifier for the PatchSet. It is used to refer to a PatchSet from within the list of Patches on a resource.
                type: string
              # Defined on PatchSetDefinition
              patches:
                # ...
              toFieldPath:
                type: string
              transforms:
                # ...
              type:
                default: FromCompositeFieldPath
                description: Type sets the patching behaviour to be used. Each patch type may require its' own fields to be set on the Patch object.
                enum:
                - FromCompositeFieldPath
                - PatchSet
                type: string
            required:
            - name
            - patches
```

Which won't work because for one, it defines `patches` as a required field (because it is, for a `PatchSetDefinition`).

I can't tell if I'm missing something and it's possible to 'hide' a field like `patches` from the CRD (i.e. make it internal) or I'm overthinking this and I should just add a `patches` field to the `PatchSet` definition (the one that refers to a `PatchSetDefinition` which in theory could be set by the user, but would be overwritten if they named a valid patchset to refer to.

Thoughts appreciated 🙂